src/include/gurt/common.h

```diff
@@ -65,14 +65,66 @@ extern "C" {
 
 /* memory allocating macros */
 void  d_free(void *);
-void *d_calloc(size_t, size_t);
-void *d_malloc(size_t);
-void *d_realloc(void *, size_t);
+void *d_calloc(int tag, size_t, size_t);
+void *d_malloc(int tag, size_t);
+void *d_realloc(int tag, void *, size_t);
 char *d_strndup(const char *s, size_t n);
 int d_asprintf(char **strp, const char *fmt, ...);
-void *d_aligned_alloc(size_t alignment, size_t size);
+void *d_aligned_alloc(int tag, size_t alignment, size_t size);
 char *d_realpath(const char *path, char *resolved_path);
 
+/** Allocation tag, please make sure dm_counters are declared in the same order. */
+enum dm_tag {
+	M_OTHER,		/* all the untagged */
+	M_STR,			/* strdup, asprintf... */
+	M_RANK,			/* rank list allocation */
+	M_CRT,			/* default cart tag */
+	M_CRT_RPC,		/* CaRT RPC */
+	M_CRT_IV,		/* IV framework in cart */
+	M_SWIM,			/* SWIM */
+	M_CSUM,			/* checksum */
+	M_PROP,			/* properties */
+	M_TSE,			/* task engine */
+	M_PL,			/* placement */
+	M_IV,			/* DAOS level IV */
+	M_SCHED,		/* ABT scheduler */
+	M_IO,			/* common I/O code */
+	M_IO_ARG,		/* SGL, IOV... */
+	M_BIO,			/* BIO */
+	M_VEA,			/* VEA */
+	M_AGG,			/* common aggregation */
+	M_RECOV,		/* common rebuild */
+	M_EC,			/* common EC */
+	M_EC_AGG,		/* EC aggregation */
+	M_EC_RECOV,		/* EC recovery */
+	M_POOL,			/* Pool, cache or data structures */
+	M_CONT,			/* container, cache or data structures */
+	M_OBJ,			/* object, shard, cache... */
+	M_DTX,			/* distributed transaction */
+	M_VOS,			/* common VOS */
+	M_VOS_TS,		/* VOS timestamp cache */
+	M_VOS_LRU,		/* VOS LRU */
+	M_VOS_DTX,		/* DTX in VOS */
+	M_RDB,			/* Raft DB */
+	M_TAG_MAX,		/* the last */
+};
+
+struct dm_tls_counter;
+
+/* 32 bytes header in memory-debugging mode */
+struct dm_header {
+	struct dm_tls_counter	*mh_counter;
+	/** real address of the buffer */
+	void			*mh_addr;
+	int64_t			 mh_size;
+	int32_t			 mh_alignment;
+	int32_t			 mh_tag;
+	uint64_t		 mh_payload[0];
```

Comment:
we can store a magic number in header, but adding a line number isn't help right? unless we have a separate counter for each line, which is unrealistic.
I thought about adding a tailer, but we already consumed 32 extra bytes everywhere, it means we have to add at least 16 bytes as tell, which is probably too much because this is enabled for dev/debug all the time (in CI)

Reply:
oh, line number is only for overflow detection, so we know which allocation is being overwritten.

I do not think extra 4 bytes(32 bit magic) at the end  will cause too much issues here, since it already allocate extra 32 bytes  for header?